### PR TITLE
WIP: JDK-8227425: Add support for e-paper displays on i.MX6 devices

### PR DESCRIPTION
**Pull request by jgneff**
My preliminary tests indicate that this pull request allows JavaFX to support the newer Kobo Glo HD and Kobo Clara HD e-readers of Issue #521 without causing any regression errors for the Kobo Touch B and Kobo Touch C e-readers of Issue #339.

To my surprise, the Kobo Clara HD is backward compatible and works with the support already added to JavaFX by Pull Request #369. The changes in the first commit of this pull request (0dc43591ceaf3ba07285b0741c630ad402d32f22) are due primarily to two problems on the Kobo Glo HD: a minor incompatibility in one of the IOCTL calls, and a bug in the reported size of the frame buffer's virtual resolution.

I marked this pull request as a *Work In Progress* so that I can do some additional testing on all of the devices, along with a comparison of their relative performance.

### CODE REVIEW

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/EPDFrameBuffer.java [hunk 1]
```diff
@@ -571,20 +586,57 @@ ByteBuffer getOffscreenBuffer() {
          * "QuantumRenderer modifies buffer in use by JavaFX Application Thread"
          * <https://bugs.openjdk.java.net/browse/JDK-8201567>.
          */
-        int size = xresVirtual * yresVirtual * Integer.SIZE;
+        int size = xresVirtual * yres * Integer.BYTES;
```
Comment (jgneff) [inline modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/EPDFrameBuffer.java:589]: This is the only problem fixed by this commit that also exists in JavaFX 13. It results in an off-screen buffer that is eight times larger than necessary when using an 8-bit or 16-bit frame buffer. The fix is to use the number of bytes per pixel (`Integer.BYTES`), not the number of bits per pixel (`Integer.SIZE`), when calculating the capacity of the 32-bit off-screen byte buffer.

File: modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/EPDFrameBuffer.java [hunk 2]
```diff
@@ -617,6 +675,14 @@ long getNativeHandle() {
      * Gets the virtual horizontal resolution of the frame buffer. See the notes
      * for the {@linkplain EPDFrameBuffer#EPDFrameBuffer constructor} above.
      *
+     * @implNote When using an 8-bit frame buffer, set by the system property
+     * {@code monocle.epd.bitsPerPixel=8}, the Kobo Touch Models N905B and N905C
+     * and the Kobo Glo HD Model N437 work only when this method returns
+     * {@code xresVirtual}, while the Kobo Clara HD Model N249 works only when
+     * this method returns {@code xres}.
+     * <p>
+     * TODO: Make 8-bit frame buffers work for all devices.</p>
```
Comment (jgneff) [inline modules/javafx.graphics/src/main/java/com/sun/glass/ui/monocle/EPDFrameBuffer.java:684]: One way to work around the bogus frame buffer resolution might be to add a system property that overrides the reported value, such as `-Dmonocle.epd.xresVirtual=1448`. Oddly enough, the error does not occur on the Kobo Clara HD when the 8-bit frame buffer is inverted with `-Dmonocle.epd.bitsPerPixel=8 -Dmonocle.epd.y8inverted=true`.

**Comment by kevinrushforth** [comment]
As announced in [this message](https://mail.openjdk.java.net/pipermail/openjfx-dev/2019-September/023551.html), the official `jfx` repository is moving from [hg.openjdk.java.net/openjfx/jfx-dev/rt](https://hg.openjdk.java.net/openjfx/jfx-dev/rt) to [github.com/openjdk/jfx](https://github.com/openjdk/jfx).

This [sandbox repository](https://github.com/javafxports/openjdk-jfx) is being retired on 1-Oct-2019. You will need to migrate your WIP pull request to the openjdk/jfx repo if you want to continue working on it. Here are [instructions for migrating your pull request](https://mail.openjdk.java.net/pipermail/openjfx-dev/2019-September/023560.html). The updated [CONTRIBUTING.md](https://github.com/openjdk/jfx/blob/master/CONTRIBUTING.md) doc has additional information on how to proceed.

Once you have done this, it would be helpful to add a comment with a pointer to the new PR.

**Comment by jgneff** [comment]
Thanks, Kevin. This pull request is almost done, so I plan to move it over to the new repository when it is ready for review and no longer a _work in progress_. The only other change I might make is to allow for working around the [bogus frame buffer value](#pullrequestreview-260458110) from one of the device models. I could also put the one-line [JavaFX 13 bug fix](#pullrequestreview-260456893) into its own bug report, but JavaFX 13 is not a target for back-porting fixes, so it's probably not worth making it separate.

**Comment by kevinrushforth** [comment]
Sounds like a plan.

**Comment by jgneff** [comment]
This pull request is now [Pull Request #60](https://github.com/openjdk/jfx/pull/60) in the new [JavaFX repository](https://github.com/openjdk/jfx).

**Pull request closed without merging**